android-libs/src/main/java/io/ooni/mk/MKResourcesManager.java

```diff
@@ -0,0 +1,118 @@
+        String logMsg = "Copy resource " + filename + " to " + destPath;
+        Log.v(LOG_TAG, logMsg + "...");
+        try {
+            copy(context.getAssets().open(filename), destPath);
+        } catch (java.io.IOException err) {
+            Log.e(LOG_TAG, logMsg, err);
+            return false; // something bad happened
+        }
+        Log.v(LOG_TAG, logMsg + "... done");
+        return true; // all good
+    }
+
+    static boolean copyResources(@NonNull Context context) {
+        if (!copySingleResource(context, ASN_DB)) {
+            return false; // cannot copy
+        }
+        if (!copySingleResource(context, CA_BUNDLE)) {
+            return false; // cannot copy
+        }
+        if (!copySingleResource(context, COUNTRY_DB)) {
+            return false; // cannot copy
+        }
+        if (!copySingleResource(context, VERSION_JSON)) {
+            return false; // cannot copy
+        }
+        return true; // all good
+    }
+
+    static byte[] readAll(@NonNull InputStream inputstream) throws java.io.IOException {
+        return org.apache.commons.io.IOUtils.toByteArray(inputstream);
+    }
+
+    static boolean weNeedToCopy(@NonNull Context context) {
+        try {
+            byte[] bundleVersion;
+            byte[] fsVersion;
+            // Implementation note: we assume that the version.json file is small
+            bundleVersion = readAll(context.getAssets().open(VERSION_JSON));
+            fsVersion = readAll(new FileInputStream(getVersionJSONPath(context)));
+            if (java.util.Arrays.equals(bundleVersion, fsVersion)) {
+                return false; // Nope
+            }
+        } catch (java.io.IOException err) {
+            Log.e(LOG_TAG, "Cannot verify whether we need to update resources", err);
+            // FALLTHROUGH
+        }
+        return true; // Yep
+    }
+
+    // maybeUpdateResources ensures that the resources copied in the storage accessible by C/C++
+    // code are current, and otherwise makes them current. The current implementation will make
+    // the resources current by copying them from the assets folder. The mechanism used to decide
+    // whether we need to copy resources over relies on a small file named version.json that is
+    // keeping tracks of the assets version. Whenever the copy of this file saved into the storage
+    // is different from the resources in the AAR, we copy resources over. Every new release of
+    // this AAR will possibly cause a bump in the resources version. This function returns true
+    // if all is good and false in case of error. See the logcat in such case.
+    public static boolean maybeUpdateResources(@NonNull Context context) {
+        if (!weNeedToCopy(context)) {
+            return true; // No error occurred
+        }
+        return copyResources(context);
+    }
+
+    @NonNull static String getPath(@NonNull Context context, @NonNull String string) {
+        return context.getFilesDir() + "/" + string;
```

Comment:
At the moment we use `getCacheDir()` instead of `getFilesDir()` so the app can free these resources when not needed and copy them again when app is restarted. For what I remember is the best approach.

Reply:
Also maybe that "/" is not needed.
From current app: `File f = new File(getCacheDir(), filename);`

Reply:
Sure, I'll change the code to use `getCacheDir`. I can also change the code to use `File`.

Reply:
New code uses the cache directory and uses File.